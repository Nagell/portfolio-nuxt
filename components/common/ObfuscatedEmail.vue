<template>
    <NuxtLink
        href="#"
        aria-label="E-Mail address"
        class="obfuscated-email hover:text-primary focus:text-primary transition-colors duration-300"
        data-name="dawidnitka"
        data-domain="gmail"
        data-tld="com"
        rel="noopener noreferrer"
        @click="openEmail"
    />
</template>

<script lang="ts" setup>
    /** Open the default mail client using the obfuscated email as the recipient. */
    function openEmail(event: MouseEvent) {
        const email = event.target as HTMLElement
        const name = email.getAttribute('data-name')
        const domain = email.getAttribute('data-domain')
        const tld = email.getAttribute('data-tld')

        window.open(`mailto:${name}@${domain}.${tld}`, '_blank')
    }
</script>

<style scoped>
    .obfuscated-email:after {
        content: attr(data-name) "@" attr(data-domain) "." attr(data-tld);
    }
</style>
