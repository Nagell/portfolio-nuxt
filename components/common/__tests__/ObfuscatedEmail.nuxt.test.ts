import { mountSuspended } from '@nuxt/test-utils/runtime'
import { describe, expect, it, vi } from 'vitest'

import ObfuscatedEmail from '../ObfuscatedEmail.vue'

describe('ObfuscatedEmail Component', () => {
    it('renders correctly', async () => {
        const wrapper = await mountSuspended(ObfuscatedEmail)

        expect(wrapper.html()).toMatchSnapshot()
    })

    it('contains all data attributes necessary for creation of an email', async () => {
        const wrapper = await mountSuspended(ObfuscatedEmail)

        expect(wrapper.attributes()['data-name']).toBeDefined()
        expect(wrapper.attributes()['data-domain']).toBeDefined()
        expect(wrapper.attributes()['data-tld']).toBeDefined()
    })

    it('on click calls the window open with a "mailto"', async () => {
        const wrapper = await mountSuspended(ObfuscatedEmail)
        const windowOpenSpy = vi.spyOn(window, 'open')

        const name = wrapper.attributes()['data-name']
        const domain = wrapper.attributes()['data-domain']
        const tld = wrapper.attributes()['data-tld']
        const expectedCall = `mailto:${name}@${domain}.${tld}`

        wrapper.trigger('click')

        expect(windowOpenSpy).toHaveBeenCalledWith(
            expectedCall,
            '_blank'
        )
    })
})
