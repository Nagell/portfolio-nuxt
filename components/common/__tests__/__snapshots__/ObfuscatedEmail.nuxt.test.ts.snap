// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ObfuscatedEmail > renders correctly 1`] = `"<a data-v-fa54199c="" href="/#" rel="noopener noreferrer" aria-label="E-Mail address" class="obfuscated-email hover:text-primary focus:text-primary transition-colors duration-300" data-name="contact" data-domain="dawidnitka" data-tld="com"></a>"`;
